File: test/windows/AudioTranscriptionSummary.Tests/ErrorLoggerTests.cs
// ErrorLoggerTests.cs
// ErrorLogger のテスト
// - 元ファイル名ベースのエラーログ出力
// - ファイル名なしの場合は app.error.log に出力
// - 既存ファイルへの追記
// - ログ内容にデバッグ情報（日時、操作、ファイル名等）が含まれること
//
// テストデータ: test/data/test.wav, test/data/test.transcript.txt
// テスト生成ファイルは test/data/ 以下に出力し、テスト完了後に削除する

using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using Xunit;

namespace AudioTranscriptionSummary.Tests;

public class ErrorLoggerTests : IDisposable
{
    private static readonly string TestDataDir = Path.GetFullPath(
        Path.Combine(AppContext.BaseDirectory, "..", "..", "..", "..", "..", "..", "..", "test", "data"));

    private readonly List<string> _generatedFiles = new();

    public void Dispose()
    {
        foreach (var file in _generatedFiles)
        {
            if (File.Exists(file))
                File.Delete(file);
        }
    }

    private void WriteTestErrorLog(Exception error, string operation, string? sourceFileName)
    {
        if (!Directory.Exists(TestDataDir))
            Directory.CreateDirectory(TestDataDir);

        string logFileName;
        if (!string.IsNullOrEmpty(sourceFileName))
        {
            var baseName = Path.GetFileNameWithoutExtension(sourceFileName);
            logFileName = $"{baseName}.error.log";
        }
        else
        {
            logFileName = "app.error.log";
        }

        var filePath = Path.Combine(TestDataDir, logFileName);
        _generatedFiles.Add(filePath);

        var sb = new StringBuilder();
        sb.AppendLine();
        sb.AppendLine("=== エラーレポート ===");
        sb.AppendLine($"日時: {DateTime.Now:yyyy-MM-dd HH:mm:ss.fff}");
        sb.AppendLine($"操作: {operation}");
        sb.AppendLine($"処理ファイル: {sourceFileName ?? "(なし)"}");
        sb.AppendLine();
        sb.AppendLine("--- エラー概要 ---");
        sb.AppendLine($"説明: {error.Message}");
        sb.AppendLine($"型: {error.GetType().FullName}");
        sb.AppendLine();
        sb.AppendLine("--- システム情報 ---");
        sb.AppendLine($"OS: {Environment.OSVersion}");
        sb.AppendLine($"PID: {Environment.ProcessId}");
        sb.AppendLine();

        File.AppendAllText(filePath, sb.ToString(), Encoding.UTF8);
    }

    [Fact]
    public void ErrorLog_WithSourceFileName_CreatesNamedLogFile()
    {
        var logFile = Path.Combine(TestDataDir, "test.error.log");
        if (File.Exists(logFile)) File.Delete(logFile);
        _generatedFiles.Add(logFile);

        var error = new InvalidOperationException("テストエラー");
        WriteTestErrorLog(error, "文字起こし", "test.wav");

        Assert.True(File.Exists(logFile));

        var content = File.ReadAllText(logFile, Encoding.UTF8);
        Assert.Contains("エラーレポート", content);
        Assert.Contains("日時:", content);
        Assert.Contains("操作: 文字起こし", content);
        Assert.Contains("処理ファイル: test.wav", content);
        Assert.Contains("テストエラー", content);
        Assert.Contains("システム情報", content);
    }

    [Fact]
    public void ErrorLog_WithoutSourceFileName_CreatesAppErrorLog()
    {
        var logFile = Path.Combine(TestDataDir, "app.error.log");
        if (File.Exists(logFile)) File.Delete(logFile);
        _generatedFiles.Add(logFile);

        var error = new Exception("ファイル名なしエラー");
        WriteTestErrorLog(error, "不明な操作", null);

        Assert.True(File.Exists(logFile));

        var content = File.ReadAllText(logFile, Encoding.UTF8);
        Assert.Contains("処理ファイル: (なし)", content);
        Assert.Contains("操作: 不明な操作", content);
    }

    [Fact]
    public void ErrorLog_AppendsToExistingFile()
    {
        var logFile = Path.Combine(TestDataDir, "append_test.error.log");
        if (File.Exists(logFile)) File.Delete(logFile);
        _generatedFiles.Add(logFile);

        WriteTestErrorLog(new Exception("エラー1"), "操作1", "append_test.wav");
        WriteTestErrorLog(new Exception("エラー2"), "操作2", "append_test.wav");

        var content = File.ReadAllText(logFile, Encoding.UTF8);
        Assert.Contains("エラー1", content);
        Assert.Contains("エラー2", content);
        Assert.Contains("操作1", content);
        Assert.Contains("操作2", content);
    }

    [Fact]
    public void TestTranscriptFileExists()
    {
        var transcriptFile = Path.Combine(TestDataDir, "test.transcript.txt");
        Assert.True(File.Exists(transcriptFile),
            "test/data/test.transcript.txt が存在しません。テストデータを配置してください。");

        var text = File.ReadAllText(transcriptFile, Encoding.UTF8);
        Assert.NotEmpty(text);
    }

    [Fact]
    public void TestAudioFileExists()
    {
        var audioFile = Path.Combine(TestDataDir, "test.m4a");
        Assert.True(File.Exists(audioFile),
            "test/data/test.m4a が存在しません。テストデータを配置してください。");
    }
}

File: macos/Tests/AudioTranscriptionSummaryTests/ErrorLoggerTests.swift
// ErrorLoggerTests.swift
// ErrorLogger のテスト
// - 元ファイル名ベースのエラーログ出力
// - ファイル名なしの場合は app.error.log に出力
// - 既存ファイルへの追記
// - ログ内容にデバッグ情報（日時、操作、ファイル名、スタックトレース等）が含まれること

import Testing
import Foundation
@testable import AudioTranscriptionSummary

/// テストデータディレクトリのパス
private let testDataDir: URL = {
    // プロジェクトルート/test/data/ を使用
    var url = URL(fileURLWithPath: #filePath)
    // macos/Tests/AudioTranscriptionSummaryTests/ → プロジェクトルート
    for _ in 0..<4 { url = url.deletingLastPathComponent() }
    return url.appendingPathComponent("test/data")
}()

@Suite("ErrorLogger テスト")
struct ErrorLoggerTests {

    /// テスト中に生成されたファイルを追跡して後で削除する
    private func cleanupFiles(_ files: [URL]) {
        for file in files {
            try? FileManager.default.removeItem(at: file)
        }
    }

    @Test("元ファイル名ベースのエラーログが生成される")
    func testErrorLogWithSourceFileName() throws {
        let logFile = testDataDir.appendingPathComponent("test.error.log")
        // 事前にクリーンアップ
        try? FileManager.default.removeItem(at: logFile)
        defer { cleanupFiles([logFile]) }

        // ErrorLogger は内部で AWSSettingsViewModel のディレクトリを使うため、
        // 直接ファイル書き込みロジックをテストする
        let error = AppError.transcriptionFailed(
            underlying: NSError(domain: "TestDomain", code: -1,
                                userInfo: [NSLocalizedDescriptionKey: "テストエラー"]))

        // テスト用にログを直接書き込む
        writeTestErrorLog(error: error, operation: "文字起こし", sourceFileName: "test.m4a", to: testDataDir)

        // ファイルが生成されたことを確認
        #expect(FileManager.default.fileExists(atPath: logFile.path))

        let content = try String(contentsOf: logFile, encoding: .utf8)
        // 必須情報が含まれていることを確認
        #expect(content.contains("エラーレポート"))
        #expect(content.contains("日時:"))
        #expect(content.contains("操作: 文字起こし"))
        #expect(content.contains("処理ファイル: test.m4a"))
        #expect(content.contains("テストエラー"))
        #expect(content.contains("システム情報"))
    }

    @Test("ファイル名なしの場合は app.error.log に出力される")
    func testErrorLogWithoutSourceFileName() throws {
        let logFile = testDataDir.appendingPathComponent("app.error.log")
        try? FileManager.default.removeItem(at: logFile)
        defer { cleanupFiles([logFile]) }

        let error = NSError(domain: "TestDomain", code: -2,
                            userInfo: [NSLocalizedDescriptionKey: "ファイル名なしエラー"])

        writeTestErrorLog(error: error, operation: "不明な操作", sourceFileName: nil, to: testDataDir)

        #expect(FileManager.default.fileExists(atPath: logFile.path))

        let content = try String(contentsOf: logFile, encoding: .utf8)
        #expect(content.contains("処理ファイル: (なし)"))
        #expect(content.contains("操作: 不明な操作"))
    }

    @Test("既存ファイルに追記される")
    func testErrorLogAppend() throws {
        let logFile = testDataDir.appendingPathComponent("append_test.error.log")
        try? FileManager.default.removeItem(at: logFile)
        defer { cleanupFiles([logFile]) }

        let error1 = NSError(domain: "Test", code: -1,
                             userInfo: [NSLocalizedDescriptionKey: "エラー1"])
        let error2 = NSError(domain: "Test", code: -2,
                             userInfo: [NSLocalizedDescriptionKey: "エラー2"])

        writeTestErrorLog(error: error1, operation: "操作1", sourceFileName: "append_test.m4a", to: testDataDir)
        writeTestErrorLog(error: error2, operation: "操作2", sourceFileName: "append_test.m4a", to: testDataDir)

        let content = try String(contentsOf: logFile, encoding: .utf8)
        // 両方のエラーが含まれていることを確認
        #expect(content.contains("エラー1"))
        #expect(content.contains("エラー2"))
        #expect(content.contains("操作1"))
        #expect(content.contains("操作2"))
    }

    @Test("テスト用文字起こしテキストが存在する")
    func testTranscriptFileExists() throws {
        let transcriptFile = testDataDir.appendingPathComponent("test.transcript.txt")
        #expect(FileManager.default.fileExists(atPath: transcriptFile.path),
                "test/data/test.transcript.txt が存在しません。テストデータを配置してください。")

        let text = try String(contentsOf: transcriptFile, encoding: .utf8)
        #expect(!text.isEmpty)
    }

    @Test("テスト用音声ファイルが存在する")
    func testAudioFileExists() {
        let audioFile = testDataDir.appendingPathComponent("test.m4a")
        #expect(FileManager.default.fileExists(atPath: audioFile.path),
                "test/data/test.m4a が存在しません。テストデータを配置してください。")
    }
}

// MARK: - テスト用ヘルパー（ErrorLogger の内部ロジックを再現）

/// ErrorLogger と同じ形式でエラーログを指定ディレクトリに書き込む
private func writeTestErrorLog(error: Error, operation: String, sourceFileName: String?, to directory: URL) {
    if !FileManager.default.fileExists(atPath: directory.path) {
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    }

    let logFileName: String
    if let name = sourceFileName, !name.isEmpty {
        let baseName = (name as NSString).deletingPathExtension
        logFileName = "\(baseName).error.log"
    } else {
        logFileName = "app.error.log"
    }
    let fileURL = directory.appendingPathComponent(logFileName)

    let fmt = DateFormatter()
    fmt.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
    fmt.locale = Locale(identifier: "en_US_POSIX")

    var lines: [String] = []
    lines.append("\n=== エラーレポート ===")
    lines.append("日時: \(fmt.string(from: Date()))")
    lines.append("操作: \(operation)")
    lines.append("処理ファイル: \(sourceFileName ?? "(なし)")")
    lines.append("")
    lines.append("--- エラー概要 ---")
    lines.append("説明: \(error.localizedDescription)")
    lines.append("型: \(type(of: error))")
    lines.append("")

    if let appErr = error as? AppError {
        lines.append("--- AppError 詳細 ---")
        lines.append("errorDescription: \(appErr.errorDescription ?? "nil")")
    }

    let nsErr = error as NSError
    lines.append("--- NSError 詳細 ---")
    lines.append("domain: \(nsErr.domain)")
    lines.append("code: \(nsErr.code)")
    lines.append("")
    lines.append("--- システム情報 ---")
    lines.append("OS: \(ProcessInfo.processInfo.operatingSystemVersionString)")
    lines.append("プロセス: \(ProcessInfo.processInfo.processName)")
    lines.append("PID: \(ProcessInfo.processInfo.processIdentifier)")

    let text = lines.joined(separator: "\n") + "\n"
    guard let data = text.data(using: .utf8) else { return }

    if FileManager.default.fileExists(atPath: fileURL.path) {
        if let handle = try? FileHandle(forWritingTo: fileURL) {
            handle.seekToEndOfFile()
            handle.write(data)
            handle.closeFile()
        }
    } else {
        try? data.write(to: fileURL)
    }
}
